pkg/registry/etcd/nse_server.go

```diff
@@ -0,0 +1,129 @@
+	"context"
+	"errors"
+	"io"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+
+	v1 "github.com/networkservicemesh/sdk-k8s/pkg/tools/k8s/apis/networkservicemesh.io/v1"
+	"github.com/networkservicemesh/sdk-k8s/pkg/tools/k8s/client/clientset/versioned"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/matchutils"
+)
+
+type etcdNSERegistryServer struct {
+	chainContext context.Context
+	client       versioned.Interface
+	ns           string
+}
+
+func (n *etcdNSERegistryServer) Register(ctx context.Context, request *registry.NetworkServiceEndpoint) (*registry.NetworkServiceEndpoint, error) {
+	resp, err := n.client.NetworkservicemeshV1().NetworkServiceEndpoints(n.ns).Create(
+		ctx,
+		&v1.NetworkServiceEndpoint{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: request.Name,
+			},
+			Spec: *(*v1.NetworkServiceEndpointSpec)(request),
+		},
+		metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	resp.Spec.DeepCopyInto((*v1.NetworkServiceEndpointSpec)(request))
+	return next.NetworkServiceEndpointRegistryServer(ctx).Register(ctx, request)
+}
+
+func (n *etcdNSERegistryServer) Find(query *registry.NetworkServiceEndpointQuery, s registry.NetworkServiceEndpointRegistry_FindServer) error {
+	if query.Watch {
+		if err := n.watch(query, s); err != nil && !errors.Is(err, io.EOF) {
+			return err
+		}
+	} else {
+		list, err := n.client.NetworkservicemeshV1().NetworkServiceEndpoints(n.ns).List(s.Context(), metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		for i := 0; i < len(list.Items); i++ {
+			item := (*registry.NetworkServiceEndpoint)(&list.Items[i].Spec)
+			if matchutils.MatchNetworkServiceEndpoints(query.NetworkServiceEndpoint, item) {
+				err := s.Send(item)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return next.NetworkServiceEndpointRegistryServer(s.Context()).Find(query, s)
+}
+
+func (n *etcdNSERegistryServer) Unregister(ctx context.Context, request *registry.NetworkServiceEndpoint) (*empty.Empty, error) {
+	err := n.client.NetworkservicemeshV1().NetworkServiceEndpoints(n.ns).Delete(
+		ctx,
+		request.Name,
+		metav1.DeleteOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return next.NetworkServiceEndpointRegistryServer(ctx).Unregister(ctx, request)
+}
+
+func (n *etcdNSERegistryServer) watch(query *registry.NetworkServiceEndpointQuery, s registry.NetworkServiceEndpointRegistry_FindServer) error {
+	watcher, err := n.client.NetworkservicemeshV1().NetworkServiceEndpoints(n.ns).Watch(s.Context(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case <-n.chainContext.Done():
+			return n.chainContext.Err()
+		case <-s.Context().Done():
+			return s.Context().Err()
+		case event := <-watcher.ResultChan():
+			if event.Type != watch.Added && event.Type != watch.Deleted {
```

Comment:
I think we need to support Updated, expect scenario:

1. I've started to watch for endpoints, recieve ep1, but it is not available du to re-start.
2. ep1 is restarted, and updated its information inside registry, so client will not be notified.

Reply:
Fixed